File: gam_TMB/gam_tmb_batches.cpp
#include <TMB.hpp>  
template<class Type>
Type objective_function<Type>::operator() (){
  using namespace density;
  using namespace Eigen; 
  
  //2D thin-plate regression spline from mgcv
  
  //Read data from R------------------
  DATA_VECTOR(y);                     // Response
  DATA_MATRIX(smoothMat);             // Design matrix (no intercept)
  DATA_SPARSE_MATRIX(penaltyMat);     // Penalty matrix
  DATA_INTEGER(penaltyDim);           // Dimensions of penalty matrix
  DATA_VECTOR(bStart);               // Start of batches
  DATA_INTEGER(bLength);              // Length of batches
  DATA_INTEGER(nbatches);             // Number of batches
  
  //Read parameters from R------------
  PARAMETER(b0);                      // Intercept
  PARAMETER_VECTOR(smoothCoefs);      // Smooth coefficients
  PARAMETER(log_lambda);              // Log penalization parameter
  PARAMETER(log_sigma);               // Log(SD) of measurement error 
  
  //Transform SD and penalization parameters 
  Type sigma = exp(log_sigma);
  Type lambda = exp(log_lambda);
  
  //Initialize objective function
  Type nll=0;
  
  // Penalization ------------------------------------
  
  SparseMatrix<Type> S = lambda*penaltyMat; //Penalty term * Penalty matrix
  nll -= 0.5*penaltyDim*log_lambda - 0.5*GMRF(S).Quadform(smoothCoefs);
  
  // Main model -------------------------------------------
  int k = 0; //Counter
  int nCoefs = smoothCoefs.size(); //Number of smoothing coefficients
  
  for(int i=0;i<nbatches;i++){ //For each batch
    matrix<Type> sMatBlock = smoothMat.block(k,0,bLength,nCoefs); //Pull out a block of model matrix starting at 0,k going for bLength rows, nCoefs columns
    vector<Type> ySeg = y.segment(k,bLength); //Pull out data for that block
    vector<Type> muSeg = b0 + sMatBlock*smoothCoefs; //Expected value for that block
    
    nll -= sum(dnorm(ySeg, muSeg, sigma, true)); //Decrement logLik
    
    k = k + bLength; //Increment counter
  }
  
  return nll; //Return likelihood
}

File: gam_TMB/gam_tmb_batches.R
library(mgcv)
library(TMB)
library(ggplot2)
library(ggpubr)
set.seed(1)

#Make 20000 sampling locations
N <- 20000
E <- runif(N,-10,10)
N <- runif(N,-10,10)
dat <- data.frame(y=0,E,N) #Fill in y with zeros for now

s0 <- gam(y~s(E,N,bs='ts',k=60),data=dat,fit=FALSE) #Create model matrix
modMat <- s0$X[,-1] #Model matrix (no intercept)
penMat <- s0$smooth[[1]]$S[[1]] #Penalty matrix
penMat <- as(penMat,'sparseMatrix') #Change to sparse matrix
penDim <- nrow(penMat) #Dimensions of penalty matrix

b0 <- 1 #Intercept
(smoothCoefs <- rnorm(ncol(modMat))) #Smoothing coefficients
# lambda <- 10 #Penalty term
# smoothCoefs <- as.vector(mvtnorm::rmvnorm(1,rep(0,nrow(penMat)),sigma=solve(penMat)/lambda)) #Draw coefs from MVnorm prior 

sigma <- 2 #Residual SD
yhat <- b0 + modMat %*% smoothCoefs #Mean values at each location
dat$y <- rnorm(N,yhat,sigma) #Generate data
dat$field <- modMat %*% smoothCoefs #Spatial field

ggplot(dat)+geom_point(aes(x=E,y=N,col=field)) #Plot spatial field (no noise)
ggplot(dat)+geom_point(aes(x=E,y=N,col=y)) #Plot data

#Input for model

#Do matrix multiplication in 5 "batches"
nbatches <- 5
batches <- rep(1:nbatches,each=nrow(dat)/nbatches)
bStart <- sapply(1:nbatches,function(x) min(which(batches==x)))-1 #Minus one because of c++ indexing
bLength <- nrow(dat)/nbatches #How long is each batch?
datList <- list(y=dat$y, smoothMat=modMat, penaltyMat=penMat, penaltyDim=penDim,bStart=bStart,bLength=bLength,nbatches=nbatches) #Data
parList <- list(b0=0, smoothCoefs=rep(0,ncol(modMat)), log_lambda=0, log_sigma=1) #Parameters

compile("gam_tmb_batches.cpp")
dyn.load(dynlib("gam_tmb_batches"))
obj <- MakeADFun(data = datList, parameters = parList, random=c('smoothCoefs'), DLL = "gam_tmb_batches")

#Memory usage spikes from 4.5GB to 16GB during Optimizing Tape stage
opt <- nlminb(obj$par,obj$fn,obj$gr) #Run model
report <- sdreport(obj)

opt$convergence==0 #Did model converge?
report #Looks OK. b0 and logsigma are very close to actual values (1 and 0.693)

#Estimates smoothing coefficients fairly well

p1 <- ggplot(data.frame(var=c('b0','logLambda','logSigma'),estimate=report$par.fixed,stdev=sqrt(diag(report$cov.fixed)),actual=c(b0,NA,log(sigma))),aes(x=var,y=estimate))+
  geom_pointrange(aes(ymax=estimate+stdev*1.96,ymin=estimate-stdev*1.96))+
  geom_point(aes(y=actual),col='red',size=2)+
  facet_wrap(~var,scales='free')+labs(x='fixed effect')+theme(axis.text.x = element_blank())

p2 <- ggplot(data.frame(predicted=report$par.random,actual=smoothCoefs))+
  geom_point(aes(x=predicted,y=actual))+geom_abline(intercept = 0,slope=1)+
  labs(title='Smoothing coefficients')

(p <- ggarrange(p1,p2,ncol=2))
